### PR TITLE
cas challenge

### PR DESCRIPTION
**Pull request by johannbarbie**
resolves https://github.com/leapdao/leap-contracts/issues/212

this will help in the review: https://docs.google.com/drawings/d/13oFjua-v_E_yaFYUbralluI-EysgtTaZb_sSgGbxG_A

### CODE REVIEW

File: contracts/PoaOperator.sol [hunk 1]
```diff
@@ -236,4 +237,107 @@ contract PoaOperator is Adminable {
...
+  }
+
+  mapping(bytes32 => mapping(uint256 => Challenge)) challenges;
+
+  function getChallenge(bytes32 _period, uint256 _slotId) public view returns (address, uint256, address) {
+    Challenge memory chal = challenges[_period][_slotId];
+    return (chal.challenger, chal.endTime, chal.slotSigner);
+  }
+
+  uint256 chalDuration;
```
Comment (troggy) [inline contracts/PoaOperator.sol]: make it public?

File: contracts/PoaOperator.sol [hunk 2]
```diff
@@ -71,10 +71,11 @@ contract PoaOperator is Adminable {
   mapping(uint256 => Slot) public slots;
 
 
-  function initialize(Bridge _bridge, Vault _vault, uint256 _epochLength) public initializer {
+  function initialize(Bridge _bridge, Vault _vault, uint256 _epochLength, uint256 _chalDuration) public initializer {
```
Comment (troggy) [inline contracts/PoaOperator.sol]: use full name `challengeDuration` here and everywhere else instead of Chal ? Allows to spend less brain cycles when reading code

File: contracts/PoaOperator.sol [hunk 3]
```diff
@@ -236,4 +237,107 @@ contract PoaOperator is Adminable {
...
+    require(msg.value == CHAL_STAKE, "invalid challenge stake");
+
+    uint256 periodTime;
+    (,periodTime,,) = bridge.periods(periodRoot);
+    // check periodRoot
+    require(periodTime > 0, "period does not exist");
+    // check slotId
+    require(_slotId < epochLength, "slotId too high");
+    // check that sig was actually 1
+    require(uint8(uint256(_casBitmap) >> (256 - _slotId)) & 0x01 == 1, "challanged sig not claimed");
```
Comment (troggy) [inline contracts/PoaOperator.sol]:
```suggestion
    require(uint8(uint256(_casBitmap) >> (256 - _slotId)) & 0x01 == 1, "challenged sig not claimed");
```

File: contracts/PoaOperator.sol [hunk 4]
```diff
@@ -236,4 +237,107 @@ contract PoaOperator is Adminable {
...
+    // check that signature matches
+    require(
+      ecrecover(_consensusRoot, _v, _r, _s) == challenges[periodRoot][_slotId].slotSigner,
+      "signature does not match"
+    );
+    // delete period Root
+
+    delete challenges[periodRoot][_slotId];
+    // dispense reward
+    require(msg.sender == _msgSender, "no frontrunning plz");
```
Comment (troggy) [inline contracts/PoaOperator.sol]: ❤️

File: test/poaOperator.js [hunk 5]
```diff
@@ -81,6 +99,71 @@ contract('PoaOperator', (accounts) => {
...
+        await operator.challengeCas(CAS, validatorRoot, consensusRoot, 1, {value: '100000000000000000'});
+
+        let challenge = await operator.getChallenge(p[3], 1);
+        assert.equal(challenge[0], accounts[0]);
+        assert.equal(challenge[2], bob);
+
+        const casRoot = merkelize(CAS, validatorRoot);
+        const vote = Tx.periodVote(1, new Input(new Outpoint(consensusRoot, 0)));
+        vote.sign(['0x7bc8feb5e1ce2927480de19d8bc1dc6874678c016ae53a2eec6a6e9df717bfac']);
+        await operator.respondCas(consensusRoot, casRoot, 1, vote.inputs[0].v, vote.inputs[0].r, vote.inputs[0].s, accounts[0]);
```
Comment (troggy) [inline test/poaOperator.js]: Two parties can call `respondCas`:
1. Slot owner itself
2. Someone else observing the chain

Slot owner has all the data in hand obviously.

Others can take consensus root and signature from the broadcasted period vote tx. Where are they supposed to take CAS bitmap from? Is it just the same submitted CAS with a challenged bit off ?

File: test/poaOperator.js [hunk 6]
```diff
@@ -81,6 +99,71 @@ contract('PoaOperator', (accounts) => {
...
+        await operator.challengeCas(CAS, validatorRoot, consensusRoot, 1, {value: '100000000000000000'});
+
+        let challenge = await operator.getChallenge(p[4], 1);
+        assert.equal(challenge[0], accounts[0]);
+        assert.equal(challenge[2], bob);
+
+        const casRoot = merkelize(CAS, validatorRoot);
+        const periodRoot = merkelize(consensusRoot, casRoot);
+
+        await operator.timeoutCas(periodRoot, 1);
```
Comment (troggy) [inline test/poaOperator.js]: should we test it with non-zero challenge duration here? First, check it fails before endTime. Then advance the time and check it succeeds now.

File: contracts/PoaOperator.sol [hunk 7]
```diff
@@ -236,4 +237,107 @@ contract PoaOperator is Adminable {
...
+
+  mapping(bytes32 => mapping(uint256 => Challenge)) challenges;
+
+  function getChallenge(bytes32 _period, uint256 _slotId) public view returns (address, uint256, address) {
+    Challenge memory chal = challenges[_period][_slotId];
+    return (chal.challenger, chal.endTime, chal.slotSigner);
+  }
+
+  uint256 chalDuration;
+  uint256 constant CHAL_STAKE = 100000000000000000; // 0.1 ETH
```
Comment (troggy) [inline contracts/PoaOperator.sol]: not sure we can initialize it like that because of Zeppellin OS. See https://docs.openzeppelin.com/sdk/2.5/writing-contracts#avoid-initial-values-in-field-declarations

Let's make it a parameter to `initialize` + setter ?

File: test/poaOperator.js (hunk 5 above)
Comment (johannbarbie) [reply test/poaOperator.js]: it is the same CAS bitmap that was submitted with the period, no changes. We only need it here to check the bit, because it is not stored anywhere when period is submitted.

File: contracts/PoaOperator.sol (hunk 7 above)
Comment (johannbarbie) [reply contracts/PoaOperator.sol]: very good point. i should probably take it from exitHandler, and save the hassle here.

File: test/poaOperator.js (hunk 6 above)
Comment (johannbarbie) [reply test/poaOperator.js]: will do

File: test/poaOperator.js (hunk 5 above)
Comment (johannbarbie) [reply test/poaOperator.js]: ideally there is an endpoint somewhere in the node, where i can pull all data about a period submission:
- slotId
- signerAddr
- CAS bitmap
- blocksRoot
- prevPeriodRoot

this information is needed in some combination for all proofs.

File: contracts/PoaOperator.sol (hunk 7 above)
Comment (johannbarbie) [reply contracts/PoaOperator.sol]: @troggy i think i addressed all

**Comment by johannbarbie** [comment]
@troggy i think i addressed all

File: contracts/PoaOperator.sol [hunk 8]
```diff
@@ -335,9 +344,11 @@ contract PoaOperator is Adminable {
...
     require(now >= chal.endTime, "time not expired yet");
     // transfer funds
-    chal.challenger.transfer(CHAL_STAKE);
+    chal.challenger.transfer(vault.exitStake());
     // delete period
     bridge.deletePeriod(_period);
   }
 
+  // solium-disable-next-line mixedcase
+  uint256[18] private ______gap;
```
Comment (troggy) [inline contracts/PoaOperator.sol]: why is this needed? 🤔

Comment (johannbarbie) [reply contracts/PoaOperator.sol]: we are using this in other contracts as well. if contract that inherit from this one have storage, and we want to add params to base contract, then this param can be used to set off the shift in storage layout.

Comment (troggy) [reply contracts/PoaOperator.sol]: aren't we already covering this in Initializable? https://github.com/leapdao/leap-contracts/blob/master/contracts/zos/Initializable.sol#L56

Comment (troggy) [reply contracts/PoaOperator.sol]: ah, 50 slots in Initializable allow us to add things to initializable, right? But we are not extending PoaOperator yet to add this gap, no?

Comment (johannbarbie) [reply contracts/PoaOperator.sol]: i'm proposing it here, so if we add storage to PosOperator later, we are not stuck. but you are right. out of scope.

Comment (troggy) [reply contracts/PoaOperator.sol]: good point. Missed the fact that PosOperator extends PoaOperator 👍

Comment (troggy) [reply contracts/PoaOperator.sol]: `challenges` mapping is not capped. Should we reserve more than 18 memory slots here? If there are a lot of challenges their storage may overlap with PosOperator, no?

Comment (johannbarbie) [reply contracts/PoaOperator.sol]: > challenges mapping is not capped

i think every variable only takes 1 position. this is not at runtime, but when the compiler assigns an address to each variable, i think.
